ext/liquid_c/vm_assembler.h

```diff
@@ -0,0 +1,106 @@
+#ifndef VM_ASSEMBLER_H
+#define VM_ASSEMBLER_H
+
+#include "c_buffer.h"
+
+enum opcode {
+    OP_LEAVE = 0,
+    OP_WRITE_RAW = 1,
+    OP_WRITE_NODE = 2,
+    OP_POP_WRITE_VARIABLE,
+    OP_PUSH_CONST,
+    OP_HASH_NEW, // rb_hash_new & rb_hash_bulk_insert
+    OP_FILTER,
+    OP_PUSH_EVAL_EXPR,
+    OP_RENDER_VARIABLE_RESCUE,
+};
+
+typedef struct vm_assembler {
+    c_buffer_t instructions;
+    c_buffer_t constants;
+    size_t max_stack_size;
+    size_t stack_size;
+} vm_assembler_t;
+
+void vm_assembler_init(vm_assembler_t *code);
+void vm_assembler_free(vm_assembler_t *code);
+void vm_assembler_gc_mark(vm_assembler_t *code);
+void vm_assembler_add_write_raw(vm_assembler_t *code, const char *string, size_t size);
+void vm_assembler_add_write_node(vm_assembler_t *code, VALUE node);
+
+static inline size_t vm_assembler_alloc_memsize(const vm_assembler_t *code)
+{
+    return c_buffer_capacity(&code->instructions) + c_buffer_capacity(&code->constants);
+}
+
+static inline void vm_assembler_write_opcode(vm_assembler_t *code, enum opcode op)
+{
+    c_buffer_write(&code->instructions, &op, 1);
+}
+
+static inline void vm_assembler_write_ruby_constant(vm_assembler_t *code, VALUE constant)
+{
+    c_buffer_write(&code->constants, &constant, sizeof(VALUE));
+}
+
+static inline void vm_assembler_increment_stack_size(vm_assembler_t *code, size_t amount)
+{
+    code->stack_size += amount;
+    if (code->stack_size > code->max_stack_size)
+        code->max_stack_size = code->stack_size;
+}
+
+
+static inline void vm_assembler_add_leave(vm_assembler_t *code)
+{
+    vm_assembler_write_opcode(code, OP_LEAVE);
+}
+
+static inline void vm_assembler_remove_leave(vm_assembler_t *code)
```

Comment:
It won't remove any instruction, since some instructions are multiple bytes.  It relies on a `ensure_not_parsing` call in block.c to check that it isn't called in the middle of assembling.

Reply:
What I meant is that the name implies that it will remove only leave instructions.

Perhaps adding something like `assert(code->instructions.data_end == OP_LEAVE)`?